packages/opentelemetry-plugin-graphql/src/graphql.ts

```diff
@@ -0,0 +1,437 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// is moving to instrumentation
+import { isWrapped } from '@opentelemetry/core';
+
+import {
+  InstrumentationBase,
+  InstrumentationConfig,
+  InstrumentationNodeModuleDefinition,
+  InstrumentationNodeModuleFile,
+  safeExecuteInTheMiddle,
+} from '@opentelemetry/instrumentation';
+import { Maybe } from 'graphql/jsutils/Maybe';
+import type * as graphqlTypes from 'graphql';
+import { GraphQLFieldResolver } from 'graphql/type/definition';
+import { SpanAttributes, SpanNames } from './enum';
+import { OTEL_GRAPHQL_DATA_SYMBOL, OTEL_SPAN_SYMBOL } from './symbols';
+
+import {
+  executeFunctionWithObj,
+  executeArgumentsArray,
+  executeType,
+  parseType,
+  validateType,
+  GraphQLPluginConfig,
+  GraphQLPluginParsedConfig,
+  OtelExecutionArgs,
+  ObjectWithGraphQLData,
+  ObjectWithOtelSpan,
+  OPERATION_NOT_SUPPORTED,
+} from './types';
+import {
+  addSpanSource,
+  endSpan,
+  getOperation,
+  wrapFieldResolver,
+  wrapFields,
+} from './utils';
+
+import { VERSION } from './version';
+import * as api from '@opentelemetry/api';
+import type { PromiseOrValue } from 'graphql/jsutils/PromiseOrValue';
+
+const DEFAULT_CONFIG: GraphQLPluginConfig = {
+  mergeItems: false,
+  depth: -1,
+  allowValues: false,
+};
+
+export class GraphQLPlugin extends InstrumentationBase {
+  constructor(config: GraphQLPluginConfig & InstrumentationConfig = {}) {
+    super('graphql', VERSION, Object.assign({}, DEFAULT_CONFIG, config));
+  }
+
+  private _getConfig(): GraphQLPluginParsedConfig {
+    return this._config as GraphQLPluginParsedConfig;
+  }
+
+  setConfig(config: GraphQLPluginConfig & InstrumentationConfig = {}) {
+    this._config = Object.assign({}, DEFAULT_CONFIG, config);
+  }
+
+  protected init() {
+    const module = new InstrumentationNodeModuleDefinition<typeof graphqlTypes>(
+      'graphql',
+      ['15.*']
+    );
+    module.files.push(this._addPatchingExecute());
+    module.files.push(this._addPatchingParser());
+    module.files.push(this._addPatchingValidate());
+
+    return module;
+  }
+
+  private _addPatchingExecute(): InstrumentationNodeModuleFile<
+    typeof graphqlTypes
+  > {
+    const file = new InstrumentationNodeModuleFile<typeof graphqlTypes>(
+      'graphql/execution/execute.js',
+      // cannot make it work with appropriate type as execute function has 2
+      //types and/cannot import function but only types
+      (moduleExports: any) => {
```

Comment:
can we use `unknown` here and check manuallly that the wrapped functions exist ?